capirca/lib/fortigate.py

```diff
@@ -0,0 +1,937 @@
+      lines += [_SP * 2 + 'set srcaddr6 "all"']
+
+    lines += [f"{_SP * 2} set action {action if action == 'accept' else 'deny'}"]
+    if action == 'reject':
+      lines += [_SP * 2 + 'set send-deny-packet enable']
+
+    if services:
+      if self._NGFW_MODE == 'policy-based':
+        lines += [_SP * 2 + 'set enforce-default-app-port disable']
+      lines += [f"{_SP * 2} set service '{services}'"]
+    else:
+      if self._NGFW_MODE == 'policy-based':
+        lines += [_SP * 2 + 'set enforce-default-app-port enable']
+  
+    if self._NGFW_MODE == 'policy-based' and self._term.application_id:
+      lines  += [f"{_SP * 2} set application {' '.join(str(v) for v in sorted(self._term.application_id))}"]
+
+    lines  += [f"{_SP * 2} set schedule {schedule_name if schedule_name else 'always'}"]
+
+    if self._term.logging:
+      if self._term.logging == 'log-both':
+        lines += [_SP * 2 + 'set logtraffic all']
+        lines += [_SP * 2 + 'set logtraffic-start enable']
+      elif self._term.logging == 'disable':
+        lines += [_SP * 2 + 'set logtraffic disable']
+      else:
+        lines += [_SP * 2 + 'set logtraffic all']
+
+    return '\n'.join(lines)
+
+
+class Fortigate(aclgenerator.ACLGenerator):
+  """A Fortigate policy object."""
+
+  _PLATFORM = 'fortigate'
+  _NGFW_MODE = 'profile-based'
+  _DEFAULT_PROTOCOL = 'ALL'
+  SUFFIX = '.fcl'
+
+  def __init__(self, *args, **kwargs):
+    self._obj_container = ObjectsContainer()
+    super().__init__(*args, **kwargs)
+
+  def _BuildTokens(self):
+    """Build supported tokens for platform.
+
+  Returns:
+    tuple containing both supported tokens and sub tokens.
+  """
+    supported_tokens, supported_sub_tokens = super()._BuildTokens()
+    supported_tokens |= {'interface',
+                         'source_interface',
+                         'destination_interface',
+                         'source_address_exclude',
+                         'destination_address_exclude',
+                         'icmp_type',
+                         'icmp_code',
+                         'application_id'}
+
+    supported_sub_tokens.update({'option': {'tcp-established'},
+                                 # Warning, some of these are mapped
+                                 # differently. See _ACTION_TABLE
+                                 'action': {'accept', 'deny',
+                                            'next', 'reject',
+                                            'reject-with-tcp-rst'}})
+
+    return supported_tokens, supported_sub_tokens
+
+  def _TranslatePolicy(self, pol, exp_info):
+    """Translate Capirca pol to fortigate pol."""
+    self.fortigate_policies = []
+    current_date = datetime.datetime.utcnow().date()
+    exp_info_date = current_date + datetime.timedelta(weeks=exp_info)
+
+    term_dup_check = set()
+
+    for header, terms in pol.filters:
+      if self._PLATFORM not in header.platforms:
+        continue
+
+      # fortigate option format:
```

Comment:
This would be a good addition (in a follow-up) to the documentation in https://github.com/google/capirca/tree/master/doc/generators